Lua Scripting

Jacob-Mango opened the pull request:
This is basic LUA scripting. Currently supported is only Audio playback. You can do everything the normal Sound class does. You can currently not add sounds to the SoundManager via the script.

Support for calling functions with arguments have been added. Use the ScriptingTest project. Pressing P plays the Cherno.ogg file. Pressing and holding 1 or 2 decrease and increases the sound. 

View the AudioScript.h and AudioScript.cpp files to understand how the Sound class was made. 

The script is located in "/scripts/test.lua".

Jacob-Mango committed:
Added LUA project to Dependencies.

Jacob-Mango committed:
Added ScriptingTest layer.

Jacob-Mango committed:
Added basic LUA scripting support.

Jacob-Mango committed:
Quick commit while changing computers.

Jacob-Mango committed:
Changed to LUAJIT 2.0 (5.1 Standard) from LUA 5.3

Jacob-Mango committed:
Changing computers. Syncing.

Jacob-Mango committed:
After two days, and two different lua wrappers, I have finally gotten classes to work.

Jacob-Mango committed:
Saving current work done. Not much.

Jacob-Mango committed:
Ignore.

Jacob-Mango committed:
Optimiezed loading times a bit. Just so I can debug scripting easier.

Jacob-Mango committed:
Added basic vectors for vec2 3 & 4. Optimized loading by 1 minute more. Now 5 second load time.

Jacob-Mango committed:
Formatting. Going to finish Quaternions tomorrow and start events (fun).

Jacob-Mango committed:
Changing computer...

Jacob-Mango committed:
Formatting code for pull request.

Jacob-Mango committed:
Changing computers. Implementing better calling method.

Jacob-Mango committed:
Calling crashes with unresolved external symbol. Needs fixing. Almost ready for pull request.

Jacob-Mango committed:
Ready for pull.

Jacob-Mango committed:
Changing default project back to Test3D.

Jacob-Mango commented:
What are the conflicts? How would I be able to tell? So I can fix it.

Jacob-Mango committed:
Privated a sound thing

KevinW1998 commented:
I would HIGHLY recommend using variadic template instead of template overloads for Scripting::Call.

Jacob-Mango commented:
@KevinW1998 Ok. Will change that. The current Scripting::Call took me around ~1 hour to create so this shouldn't take long as well.

Jacob-Mango committed:
Changed to Vardiac Templates. Most of it is just copied over from Log.h and rewritten to support scripts.

Jacob-Mango commented:
Added Variadic Templates in 18 minutes. :)

Jacob-Mango commented:
What is this Continous-Intergration error about?

Jacob-Mango commented:
Also, what are the conflicts? So I can fix them.

TheCherno commented:
Hey Jacob, thanks so much for putting this together. Since this is quite a large addition, I'll have to check it out and do some proper testing/code review, and this may result in some follow up revisions. I'll try and do this as soon as possible, and get back to you with any actions required on your part.

As for the CI failure, the reason it failed was because you have merge conflicts in your current revision. This is because `master` has changed since you last merged it into your branch. You'll need to resolve these conflicts before I can test/merge them. Check out something like [this article](http://dont-be-afraid-to-commit.readthedocs.io/en/latest/git/conflicts.html) if you're not sure how to do that. :)

Gogsi commented:
Just to help you out, the conflict is in Sandbox.cpp 
In b21421a you change line 37 and 40, but these lines don't exist now.
The easiest fix I can think of is downloading the current Sandbox.cpp and pushing it on your fork.

Jacob-Mango committed:
Should merge no errors now.

Jacob-Mango committed:
Should work now...

Jacob-Mango committed:
Final commit... Code is synced. Problem still unknown.

Jacob-Mango committed:
Merge branch 'master' of https://github.com/TheCherno/Sparky into Scripting

Jacob-Mango committed:
Another commit.

Jacob-Mango committed:
And, the final commit. It should merge just fine now.

Jacob-Mango commented:
Hey :)

After a fair amount of commits, I have successfully merged the code together. While also learning more about git and how annoying it can be at sometimes.

It should all work fine.

Jacob-Mango commented:
All conflicts seemed to have been fixed. I am only going to change 2 files to use a different LUA wrapper maybe due to the current one may be hard to actually implement.

Jacob-Mango committed:
Made class creation easier for Lua. Look in AudioScript.cpp to see how it has changed.

KevinW1998 commented:
Ahh, good ol' luabind. It is one of my favourite front-end lua binders. It definitely offers a much nicer syntax.

Jacob-Mango commented:
@KevinW1998 yea. i was originally going to use it but didn't want to actually use boost. that was until i found a version that doesn't need boost.

Jacob-Mango committed:
Added most of the API for Sparky. Just need graphics and some bugs to be fixed.

Jacob-Mango commented:
Really only thing that is needed to be added is the ability to add and load materials, layers, text, textures. Basically anything to do with graphics.

Since that isn't as important, I am wondering if this can start under the testing. There is currently no pull errors.

Tell me what might need to be fixed or changed and I will do it.

@TheCherno 

E: Accidentally clicked closed, well it is reopened.

KevinW1998 commented:
> Really only thing that is needed to be added is the ability to add and load materials, layers, text, textures. 

And that's something we should have a wrapper generator for.
(It's a shame that C++ doesn't support Reflection)

Jacob-Mango commented:
@KevinW1998 I can try to make one (again). Just found this to be easier for now.

KevinW1998 commented:
Shouldn't we wait until cligen is finished and then extending it with luabind wrapper generation?

Jacob-Mango commented:
@KevinW1998 that might be a great idea. ok. i will just implement the graphics then I will create a new pull request (or you if you wish) that adds a luabind wrapper gen.

TheCherno commented:
I'm extremely busy (as always), so I'm just going to drop my thoughts here and we can discuss them. I haven't had the chance to look into this _too_ much, so take everything I say with a grain of salt.
- Firstly, Lua should be written as "Lua" rather than "LUA". I have no idea where the latter comes from, but for some reason many people decide to write it incorrectly
- I would like to have Lua's source code in the actual Sparky solution, rather than just as a binary
- I want to be able to use the _latest_ version of Lua
- The previous point means that we can't use LuaJIT. I'm okay with that, and in fact I'm not exactly a huge fan of LuaJIT anyway. It doesn't work properly on many platforms/archs (yes I know it claims otherwise), and it adds extra complexity which isn't worth the benefits in my opinion. It also means that we have to use an old version of Lua. So overall, it is not worth it
- As far as wrappers go, I'm okay with _one_ wrapper/binding helper library, however let's not go overboard and use too many
- The biggest issue with the current solution we have is the binding. It's incredibly difficult to maintain and requires a separate method for each function, which is unacceptable. Furthermore, the Lua binding should be stored in the actual C++ class file of the class that is being bound, rather than in an additional file (eg. instead of AudioScript.h/.cpp, it should be in Audio.h/.cpp).

I _briefly_ took some time to come up with a better solution for binding, using templates. Here's a _rough_ version of what I came up with:

``` c++
    template<typename R>
    struct Marshal
    {
        static int Dispatch(lua_State* L, const R& value) { SP_ASSERT(false); return 0; }
    };

    template<>
    struct Marshal<float>
    {
        static int Dispatch(lua_State* L, const float& value)
        {
            lua_pushnumber(L, value);
            return 1;
        }
    };

    template<typename R>
    struct Unmarshal
    {
        static R&& Dispatch(lua_State* L, int index) { SP_ASSERT(false); return 0; }
    };

    template<>
    struct Unmarshal<float>
    {
        static float&& Dispatch(lua_State* L, int index)
        {
            return lua_tonumber(L, index);
        }
    };

    template<>
    struct Unmarshal<String>
    {
        static String&& Dispatch(lua_State* L, int index)
        {
            return lua_tostring(L, index);
        }
    };

    template<typename T>
    struct FunctionWrapper
    {

    };

    template<typename T, typename R, typename... Args>
    struct FunctionWrapper<R(T::*)(Args...)>
    {
        enum { ArgCount = sizeof...(Args) };

        typedef R(T::* FunctionType)(Args...);

        template<FunctionType method, size_t... index>
        static lua_CFunction WrapMethod()
        {
            return [](lua_State *L)
            {
                T* obj = luaW_check<T>(L, 1);
                Marshal<std::remove_const<R>::type>::Dispatch(L, (obj->*method)(Unmarshal<Args>::Dispatch(L, index + 2)...));
                return 1;
            };
        }

        template<FunctionType method, size_t... index>
        static lua_CFunction WrapMethod(std::index_sequence<index...>)
        {
            return WrapMethod<method, index...>();
        }
    };

    template<typename T, typename R, typename... Args>
    struct FunctionWrapper<R(T::*)(Args...) const>
    {
        enum { ArgCount = sizeof...(Args) };

        typedef R(T::* FunctionType)(Args...) const;
        typedef R(T::* FunctionTypeNonConst)(Args...);
        template <FunctionType method, size_t... index> static lua_CFunction WrapMethod()
        {
            return FunctionWrapper<FunctionTypeNonConst>::WrapMethod<(FunctionTypeNonConst)method, index...>();
        }
        template <FunctionType method, size_t... index> static lua_CFunction WrapMethodInternal(std::index_sequence<index...>)
        {
            return FunctionWrapper<FunctionTypeNonConst>::WrapMethod<(FunctionTypeNonConst)method, index...>();
        }
    };

    template<typename T, typename... Args>
    struct FunctionWrapper<void(T::*)(Args...)>
    {
        enum { ArgCount = sizeof...(Args) };
        typedef void (T::* FunctionType)(Args...);

        template<FunctionType method, size_t... index>
        static lua_CFunction WrapMethod()
        {
            return [](lua_State *L)
            {
                T* obj = luaW_check<T>(L, 1);
                (obj->*method)(Unmarshal<Args>::Dispatch(L, index + 2)...);
                return 0;
            };
        }

        template<FunctionType method, size_t... index>
        static lua_CFunction WrapMethodInternal(std::index_sequence<index...>)
        {
            return WrapMethod<method, index...>();
        }
    };

    template<typename T, T method, std::size_t N, typename Indices = std::make_index_sequence<N>>
    lua_CFunction WrapMethod()
    {
        return FunctionWrapper<T>::WrapMethodInternal<method>(Indices());
    }

    template<typename T, T method>
    lua_CFunction WrapMethod()
    {
        return WrapMethod<T, method, FunctionWrapper<T>::ArgCount>();
    }

    #define SCRIPT_METHOD(S) WrapMethod<decltype(&S), &S>()

    Sound* L_Sound_New(lua_State *L)
    {
        String name = lua_tostring(L, 1);
        String filename = lua_tostring(L, 2);
        return spnew Sound(name, filename);
    }

    int L_Register_Audio_Classes(lua_State *L)
    {
        luaL_Reg L_Sound_Table[] =
        {
            { NULL, NULL }
        };

        luaL_Reg L_Sound_MetaTable[] =
        {
            { "Play", SCRIPT_METHOD(Sound::Play) },
            { "Loop", SCRIPT_METHOD(Sound::Loop) },
            { "Pause", SCRIPT_METHOD(Sound::Pause) },
            { "Stop", SCRIPT_METHOD(Sound::Stop) },
            { "Resume", SCRIPT_METHOD(Sound::Resume) },
            { "GetGain", SCRIPT_METHOD(Sound::GetGain) },
            { "SetGain", SCRIPT_METHOD(Sound::SetGain) },
            { NULL, NULL }
        };
        luaW_register<Sound>(L, "Sound", L_Sound_Table, L_Sound_MetaTable, L_Sound_New);
        return 1;
    }
```

Hopefully you get the idea of what a _good_ solution would look like, however this obviously isn't complete yet (i.e. constructors aren't supported yet).

Anyway, leave your thoughts here; let's talk about this.

KevinW1998 commented:
> -  I want to be able to use the latest version of Lua
> - The previous point means that we can't use LuaJIT. I'm okay with that, and in fact I'm not exactly a huge fan of LuaJIT anyway. It doesn't work properly on many platforms/archs (yes I know it claims otherwise), and it adds extra complexity which isn't worth the benefits in my opinion. It also means that we have to use an old version of Lua. So overall, it is not worth it

Well, I had never really issues with LuaJIT (But maybe because I only used it on Windows). 

> - The biggest issue with the current solution we have is the binding. It's incredibly difficult to maintain and requires a separate method for each function, which is unacceptable. Furthermore, the Lua binding should be stored in the actual C++ class file of the class that is being bound, rather than in an additional file (eg. instead of AudioScript.h/.cpp, it should be in Audio.h/.cpp).

I still suggest using a binding generator (like cligen, but for lua bindings).

TheCherno commented:
I'm definitely all for a binding generator, but for that to work there still needs to be the base binding code, such as my example above.

Jacob-Mango commented:
> As far as wrappers go, I'm okay with one wrapper/binding helper library, however let's not go overboard and use too many

Well, that was the plan. The easiest binding to use thus far is LuaBind which is what the project is using. It is only one wrapper and no more wrappers are needed.

> The biggest issue with the current solution we have is the binding. It's incredibly difficult to maintain and requires a separate method for each function, which is unacceptable. 

Noticed you were using an outdated pull. I have changed wrappers to LuaBind. Currently using a version not using boost although there is a boost version which has (some little) additional features. it seems fairly easy to use and so far no problems. 

> Furthermore, the Lua binding should be stored in the actual C++ class file of the class that is being bound, rather than in an additional file (eg. instead of AudioScript.h/.cpp, it should be in Audio.h/.cpp).

We could do two things here. We can create a generator which reads every file and parses it into one giant .h/.cpp file or we can go through each file and create a module for each. Currently, as of latest commit, it is mostly in one file,non generated. 

> there still needs to be the base binding code, such as my example above.

LuaBind seems to be working well for that. It is readable and easy enough so if a generator was to be made, no additional binding code would be needed.

I will merge LuaBind into the project next commit.

Jacob

Jacob-Mango committed:
Putting LuaBind in project.

Jacob-Mango commented:
Just wondering if you are thinking about letting us see the CI logs? Seeing it fail yet other pull requests not failing makes me wonder what is the problem.

Gogsi commented:
You can actually see the error if you click on details and then either debug or release.
In this case it's 
`LINK : fatal error LNK1104: cannot open file 'luabind-deboostified.lib' [C:\projects\sparky\Sparky-core\Sparky-core.vcxproj]
`

KevinW1998 commented:
Yea, I think you have to set luabind as dependency so it is build first

Jacob-Mango committed:
CI fix most likely.

Jacob-Mango committed:
Trying to allow classes to also be called. Currently can't be called (never actually could apparently).

Jacob-Mango committed:
Trying to allow classes to also be called. Currently can't be called (never actually could apparently).

Jacob-Mango committed:
Merge branch 'Scripting' of https://github.com/Jacob-Mango/Sparky.git

Jacob-Mango committed:
Seperated LuaBind again... (Sorry)

18 unresolved eexternal symbol errors which I am too tired to fix are in this commit.

Jacob-Mango committed:
Calling now works better. Passing in classes still has some problems.

Jacob-Mango commented:
Git did something wierd... It deleted a file which wasnt there...

Jacob-Mango committed:
Appendment of Previous Commit

Jacob-Mango committed:
Release CI fix. (Still that Class calling bug)

Jacob-Mango committed:
I can not fix commits can I?

Jacob-Mango committed:
Couple more things to do then it is finished.

Jacob-Mango committed:
I think I am rusty on my Lua.

Jacob-Mango committed:
Appendment. Lua fixes.

Jacob-Mango committed:
Unsure what that build error is. Will fix tomorrow.

Jacob-Mango commented:
Hello @TheCherno,
I have been looking into implementing Lua scripting your way and so far, it has been well. I have decided to first implement in a different project (a new one) so I can easily figure if it may be bug with your code or mine easier and so I won't modify files I do not need to. This project can be found at;

https://github.com/Jacob-Mango/MangoLua (Couldn't think of a name)

The amount of work I will be doing will be decreased during these next few weeks due to exams. Sorry.

Jacob-Mango committed:
Now using MangoLua (Modified LuaWrapper).
There seems to be a problem with static functions.

Jacob-Mango committed:
Revert "I think I am rusty on my Lua."

This reverts commit fdc2e5c9ede500aff2d6b1ee002cb955da32db10.

Jacob-Mango committed:
Simplified Lua support. Added return in calling Lua created functions (causes crash right now).

Jacob-Mango committed:
Fixing Git Ignore

Jacob-Mango committed:
Some weird crash on sound loading. Gonna trace back.

Jacob-Mango committed:
Fixed the sound bug. Added a FunctionCaller.

Jacob-Mango committed:
Merge branch 'Scripting' of https://github.com/Jacob-Mango/Sparky.git

Jacob-Mango committed:
Just the nightly push of what I have done today.

Adding a generator.

Jacob-Mango committed:
Added ScriptGen project. Based of CLIGen branch. Same bug as in CLIGen branch where it won't fully execute due to Clang "Wrong kind of initialization expression!"

Jacob-Mango commented:
Alright. Fixed the ScriptGen compile bug. Turns out you have to use LLVM 3.4.1 (I am using windows binaries).

Jacob-Mango committed:
The script generator (ScriptGen) now does Constructors and outputs into a file.

Jacob-Mango committed:
Well... it doesn't work. I can say that. The scritping generator does though. So theirs that.

* Added PUBLIC, PRIVATE and PROTECTED access types for the clang parser.
* Prints out all public (non-static) contructors, classes and functions.

Jacob-Mango commented:
Problem. Function overloading not working.

Jacob-Mango committed:
Nightly push.

Jacob-Mango committed:
Improvements. Kinda.

. Unresolved externals added
. Luabind added (again... I should make up my mind. End result is what matters right?)
. ScriptGen needs to sort between what is allowed and isn't.

Jacob-Mango committed:
Night push

Jacob-Mango committed:
Another forgetful push and commit.

All that is left is for the writer to work better it seems.

Jacob-Mango committed:
Nightly push.

Jacob-Mango committed:
Made a bug.

Jacob-Mango committed:
Night commit.

. Bug fixes.

KevinW1998 commented on Sparky-core/src/sp/scripting/script.h:
```diff
@@ -30,7 +30,16 @@ namespace sp { namespace scripting {
+			{
+				SP_ERROR("Lua >> ", lua_tostring(m_State, -1));
+			}
+			
```
Personally I wouldn't recommend surpressing the exception. Either notify the caller about an error with a bool value or throw an own exception.

Jacob-Mango commented:
That was just for debugging. Forgot to remove that. 

All that is left from that I see visible is to finish the generator and then it should all work. 

Tell me if you see something missing.

KevinW1998 commented:
Well there are some recommendation I would give:
- script.h @ line 35 should call std::forward<br>
  instead of: <br>
  `luabind::call_function<void>(m_State, func, args...);`<br>
  we can use perfect forwarding:<br>
  `luabind::call_function<void>(m_State, func, std::forward<Args>(args)...);`
- script.h @ line 24: I personally recommend using std::vector< std::string > instead of std::vector< const char\* >, because this will ensure that the string will be valid through the lifetime of the script object. 
- script.cpp @ line 8: Why do you push an empty string?
- In addition I would recommend closing the lua state on destruction of the Script object for proper cleanup

Jacob-Mango commented:
Thank you for the recomendations
- I will use forwarding. I dont know why I didn't considering I did use it last time.
- I had a problem with using a string vector. It was to do with the instance of the string being deleted as it was being pushed into the string. I changed to chars and the problem fixed itself. I will look into why strings didn't work. 
- Was one of the things I was doing to try and fix the above problem. Forgot to remove that line.

KevinW1998 commented:
>  It was to do with the instance of the string being deleted as it was being pushed into the string.

This sounds very wrong. If you still have issues with it, then I'll try it.

Jacob-Mango commented:
Sorry but I feel that I can not improve this. I will close this pull request now. Feel free to use what I have done.
